tools/iree-tracy-print.cc

```diff
@@ -0,0 +1,304 @@
+#include "third_party/tracy/server/TracyFileRead.hpp"
+#include "third_party/tracy/server/TracyFileWrite.hpp"
+#include "third_party/tracy/server/TracyWorker.hpp"
+
+IREE_FLAG(string, input_tracy_file, "",
+          "Tracy file to read. Ignored if executable command is given.");
+IREE_FLAG(string, output_tracy_file, "",
+          "Tracy file to write as the output of the given executable command.");
+IREE_FLAG(string, output_xplane_file, "",
+          "Xplane file to write as the output of execution or conversion.");
+IREE_FLAG(int32_t, tracy_port, 18086, "TCP port number for tracy profiler.");
+
+namespace {
+
+// Flag to inform main thread to disconnect from tracy profiler on SIGINT.
+std::atomic<bool> g_disconnect = false;
+
+void HandleSigInt(int sigid) {
+  g_disconnect = true;
+}
+
+// Runs a subprogram with |argv| which is null-terminated at the end.
+int RunExecutable(char** argv) {
+  int pid = fork();
+  if (pid == 0) {
+    auto tracy_port_str = std::to_string(FLAG_tracy_port);
+    setenv("TRACY_NO_EXIT", "1", /*overwrite=*/1);
+    setenv("TRACY_PORT", tracy_port_str.c_str(), /*overwrite=*/1);
+    execv(argv[0], argv);
+    exit(errno);  // Not expected to be reached.
+  }
+
+  if (pid < 0) {
+    return -1;
+  }
+
+  // Wait a little bit to check if a child process is still running.
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  int wstatus;
+  int ret = waitpid(pid, &wstatus, WNOHANG);
+  if (ret == 0) {
+    return 0;
+  }
+
+  if (ret > 0) {
+    if (WIFSIGNALED(wstatus)) {
+      errno = EINTR;
+    } else {
+      errno = WEXITSTATUS(wstatus);
+    }
+  }
+  return -1;
+}
+
+template <typename T>
+struct Zone {
+  int16_t source_location_id;
+  const T* zone;
+};
+
+template <typename T>
+std::vector<Zone<T>> GetTopZones(
+    const tracy::unordered_flat_map<int16_t, T>& zones,
+    int num_zones_to_return) {
+  std::map<int, Zone<T>> ordered_map;
+  for (const auto& z : zones) {
+    ordered_map[z.second.total] = Zone<T>{z.first, &z.second};
+    if (ordered_map.size() > num_zones_to_return) {
+      ordered_map.erase(ordered_map.begin());
+    }
+  }
+
+  std::vector<Zone<T>> vector_to_return;
+  for (auto it = ordered_map.rbegin(); it != ordered_map.rend(); ++it) {
+    vector_to_return.push_back(it->second);
+  }
+  return vector_to_return;
+};
+
+const char* GetZoneName(const tracy::Worker& worker,
+                        int16_t source_location_id) {
+  const auto& srcloc = worker.GetSourceLocation(source_location_id);
+  return worker.GetString(srcloc.name.active ? srcloc.name : srcloc.function);
+}
+
+const char* ArchToString(tracy::CpuArchitecture arch) {
+  switch (arch) {
+    case tracy::CpuArchUnknown: return "Unknown";
+    case tracy::CpuArchX86: return "x86";
+    case tracy::CpuArchX64: return "x86_64";
+    case tracy::CpuArchArm32: return "arm";
+    case tracy::CpuArchArm64: return "aarch64";
+    default: return "Unknown";
+  }
+}
+
+void PrintWorker(const tracy::Worker& worker) {
```

Comment:
I think we need a separate file for each output type, i.e., print and xplane. This would help us create a separate converting tool like `iree-tracy-convert`, where we can reuse the files.

Reply:
Done